### DISCUSSION TITLE
How do stub files work and how to ensure that they're used?

### DISCUSSION THREAD

**Discussion by songololo** [microsoft/pyright#3463] [Q&A] [answered]
I'm trying to create a stub file for `networkx` "MultiGraphs".

I have a `<package>/typings` directory and this is set inside my `pyproject.toml` file.

Stub files have been generated per these instructions: https://github.com/microsoft/pyright/blob/main/docs/type-stubs.md

I've then gone into the file of interest and tweaked the stubs, but I can't seem to get `pyright` to recognise the types (I'm using `pylance` in VSCode).

This is how my file looks:

```python
"""
This type stub file was generated by pyright.
"""

from typing import Any

from networkx.classes.graph import Graph

from networkx.classes import multigraph, multidigraph
from networkx.classes.coreviews import MultiAdjacencyView
from networkx.classes.reportviews import MultiEdgeView, MultiDegreeView

class MultiGraph(Graph):
    edge_key_dict_factory = Any
    def __init__(
        self,
        incoming_graph_data: Any | None = ...,
        multigraph_input: Any | None = ...,
        **attr
    ) -> None: ...
    @property
    def adj(self) -> MultiAdjacencyView: ...
    def new_edge_key(self, u: int, v: int) -> int: ...
    def add_edge(self, u_for_edge: int, v_for_edge: int, key=..., **attr) -> None: ...
    def add_edges_from(self, ebunch_to_add: list[tuple[int, int]], **attr) -> None: ...
    def remove_edge(self, u: int, v: int, key=...) -> None: ...
    def remove_edges_from(self, ebunch: list[tuple[int, int]]) -> None: ...
    def has_edge(self, u: int, v: int, key=...): -> bool: ...
    @property
    def edges(self) -> MultiEdgeView: ...
    def get_edge_data(self, u: int, v: int, key=..., default=...) -> None: ...
    @property
    def degree(self) -> MultiDegreeView: ...
    def is_multigraph(self) -> bool: ...
    def is_directed(self) -> bool: ...
    def copy(self, as_view=...) -> multigraph.MultiGraph: ...
    def to_directed(self, as_view=...) -> multidigraph.MultiDiGraph: ...
    def to_undirected(self, as_view=...): -> multigraph.MultiGraph: ...
    def number_of_edges(self, u: int, v: int=...) -> int: ...

```

When I import and use the networkX package, e.g.

```python
import networkx as nx

def nx_simple_geoms(nx_multigraph: nx.MultiGraph) -> None:
    pass
```

Then I get these forms of errors:
```bash
Type of parameter "nx_multigraph" is unknownPylance[reportUnknownParameterType]
Type of "MultiGraph" is unknownPylance[reportUnknownMemberType]
```

And similar for any other usage of the MultiGraph type.

Thanks.

**Comment by erictraut** [comment, marked as answer]
Did you create a subdirectory called `networkx` under `typings` and place the generated type stub files within that subdirectory?

If you right click on `nx` and choose "Go To Declaration", does it take you to the expected stub file?

**Reply by songololo** [reply]
Thanks or the fast response.

Yes, those folders have been created.

I seem to be getting tripped up somewhere along the chained module imports as part of the `nx` package.

When I click on "Go To Declaration" it takes me to `<my-package>/typings/networkx/__init__.py`.

This `__init__.py` file contains `from networkx import classes`, and this module, in turn, contains `from .multigraph import MultiGraph`.

I tried changing the imports from the above example to directly import from the file with the multigraph definition per:
```python
from networkx.classes import multigraph

def nx_simple_geoms(nx_multigraph: multigraph.MultiGraph) -> None:
    pass
```

And then it does work as intended.

**Reply by erictraut** [reply]
That means that the type information in your current stubs indicate that the top-level `networkx` module doesn't export the symbol `MultiGraph`. You would need to import that symbol from the submodule `networkx.classes`.

If you think that the `networkx` module is intended to export the `MultiGraph` symbol, you can update the stubs accordingly. You can do this by adding a redundant form of an import statement to `__init__.pyi`. By "redundant form", I mean `from .classes import MultiGraph as MultiGraph`. PEP 484 indicates that this redundant form used within a stub file should be interpreted by type checkers as a "re-export" of that symbol. If you don't use the redundant form, type checkers will assume that it's a private import — that it is not intended to be re-exported from that module.

**Reply by songololo** [reply]
Thanks! Yes, the redundant form of import statement fixes it!

Much appreciated, really enjoying the pyright dev experience so far.

**Comment by mishamsk** [comment]
Hi, was looking for networks stubs and stumbled upon this. Wondering if you've open sourced the stubs or willing to do so? Happy to contribute too! 

So far found this https://github.com/eggplants/networkx-stubs - but it seems to be abandoned and not finished
